### PR TITLE
Gracefully handle errors within node engine (instead of panic)

### PR DESCRIPTION
**Pull request by bitwiseguy**
Resolves https://github.com/statechannels/go-nitro/issues/1275

* First commit resolves https://github.com/statechannels/go-nitro/issues/1257
* Second commit resolves https://github.com/statechannels/go-nitro/issues/1263

Instead of the node engine panicking, errors are logged and are returned on the `toApi` chan as `FailedObjectives`.

### CODE REVIEW

File: node/engine/store/durablestore.go [hunk 1]
```diff
@@ -589,18 +589,19 @@ func (ds *DurableStore) SetVoucherInfo(channelId types.Destination, v payments.V
...
 }
 
-func (ds *DurableStore) GetVoucherInfo(channelId types.Destination) (v *payments.VoucherInfo, ok bool) {
+func (ds *DurableStore) GetVoucherInfo(channelId types.Destination) (*payments.VoucherInfo, error) {
+	v := &payments.VoucherInfo{}
 	err := ds.vouchers.View(func(tx *buntdb.Tx) error {
 		vJSON, err := tx.Get(channelId.String())
 		if err != nil {
-			return nil
+			return err
```
Comment (bitwiseguy) [inline node/engine/store/durablestore.go:597]: Changing this line allows us to better handle the non-existent channel [here](https://github.com/statechannels/go-nitro/blob/5ec7d46213d97cf631badd773d82bab3fcf9dfe6/payments/voucher_manager.go#L55).

File: node/engine/engine.go [hunk 2]
```diff
@@ -854,20 +850,7 @@ func (e *Engine) recordMessageMetrics(message protocols.Message) {
...
 func (e *Engine) checkError(err error) {
 	if err != nil {
-
 		e.logger.Err(err).Msgf("%s, error in run loop", e.store.GetAddress())
-
-		for _, nonFatalError := range nonFatalErrors {
-			if errors.Is(err, nonFatalError) {
-				return
-			}
-		}
```
Comment (kerzhner) [inline node/engine/engine.go:864]: In my thinking, there is a class of irrecoverable errors for which the engine should panic (at least for now, while go-nitro is under rapid development). Panicking and exiting will help us diagnose unexpected nitro implementation issues.

There is a second class of non-fatal errors. These are errors a nitro node expects to encounter, and the node can continue to operate after encountering the error. Attempting to create a second ledger channel between a pair of participants is an example of a non-fatal error.

File: node/engine/store/memstore.go [hunk 3]
```diff
@@ -472,18 +472,18 @@ func (ms *MemStore) SetVoucherInfo(channelId types.Destination, v payments.Vouch
 	return nil
 }
 
-func (ms *MemStore) GetVoucherInfo(channelId types.Destination) (v *payments.VoucherInfo, ok bool) {
+func (ms *MemStore) GetVoucherInfo(channelId types.Destination) (*payments.VoucherInfo, error) {
 	data, ok := ms.vouchers.Load(channelId.String())
 	if !ok {
-		return nil, false
+		return nil, fmt.Errorf("could not load vouchers from memory")
```
Comment (kerzhner) [inline node/engine/store/memstore.go:478]: Let's include the channelId in the error.

File: payments/payments_test.go [hunk 4]
```diff
@@ -39,8 +39,12 @@ func (svs *simpleVoucherStore) SetVoucherInfo(channelId types.Destination, v Vou
 	return nil
 }
 
-func (svs *simpleVoucherStore) GetVoucherInfo(channelId types.Destination) (v *VoucherInfo, ok bool) {
-	return svs.vouchers.Load(channelId.String())
+func (svs *simpleVoucherStore) GetVoucherInfo(channelId types.Destination) (*VoucherInfo, error) {
+	v, ok := svs.vouchers.Load(channelId.String())
+	if !ok {
+		return v, fmt.Errorf("could not load vouchers")
```
Comment (kerzhner) [inline payments/payments_test.go:45]: Let's include the channel id as part of the error.

File: payments/voucher_manager.go [hunk 5]
```diff
@@ -51,9 +51,8 @@ func (vm *VoucherManager) Remove(channelId types.Destination) error {
 // Pay will deduct amount from balance and add it to paid, returning a signed voucher for the
 // total amount paid.
 func (vm *VoucherManager) Pay(channelId types.Destination, amount *big.Int, pk []byte) (Voucher, error) {
-	vInfo, ok := vm.store.GetVoucherInfo(channelId)
-
-	if !ok {
+	vInfo, err := vm.store.GetVoucherInfo(channelId)
+	if err != nil {
 		return Voucher{}, fmt.Errorf("channel not found")
```
Comment (kerzhner) [inline payments/voucher_manager.go:56]: We should probably [wrap](https://go.dev/blog/go1.13-errors) the error returned by GetVoucherInfo when creating the new error.

File: payments/voucher_manager.go [hunk 6]
```diff
@@ -82,8 +81,8 @@ func (vm *VoucherManager) Pay(channelId types.Destination, amount *big.Int, pk [
 
 // Receive validates the incoming voucher, and returns the total amount received so far
 func (vm *VoucherManager) Receive(voucher Voucher) (*big.Int, error) {
-	vInfo, ok := vm.store.GetVoucherInfo(voucher.ChannelId)
-	if !ok {
+	vInfo, err := vm.store.GetVoucherInfo(voucher.ChannelId)
+	if err != nil {
 		return &big.Int{}, fmt.Errorf("channel not registered")
```
Comment (kerzhner) [inline payments/voucher_manager.go:86]: Another place where error wrapping is likely helpful.

File: payments/voucher_manager.go [hunk 7]
```diff
@@ -121,23 +120,23 @@ func (vm *VoucherManager) Receive(voucher Voucher) (*big.Int, error) {
...
+	return err == nil
 }
 
 // Paid returns the total amount paid so far on a channel
 func (vm *VoucherManager) Paid(chanId types.Destination) (*big.Int, error) {
-	v, ok := vm.store.GetVoucherInfo(chanId)
-	if !ok {
+	v, err := vm.store.GetVoucherInfo(chanId)
+	if err != nil {
 		return &big.Int{}, fmt.Errorf("channel not registered")
```
Comment (kerzhner) [inline payments/voucher_manager.go:131]: Let's wrap the upstream error.

File: payments/voucher_manager.go [hunk 8]
```diff
@@ -121,23 +120,23 @@ func (vm *VoucherManager) Receive(voucher Voucher) (*big.Int, error) {
...
 	return v.LargestVoucher.Amount, nil
 }
 
 // Remaining returns the remaining amount of funds in the channel
 func (vm *VoucherManager) Remaining(chanId types.Destination) (*big.Int, error) {
-	v, ok := vm.store.GetVoucherInfo(chanId)
-	if !ok {
+	v, err := vm.store.GetVoucherInfo(chanId)
+	if err != nil {
 		return &big.Int{}, fmt.Errorf("channel not registered")
```
Comment (kerzhner) [inline payments/voucher_manager.go:140]: Let's wrap the upstream error.

File: node/engine/engine.go [hunk 9]
```diff
@@ -863,11 +864,6 @@ func (e *Engine) checkError(err error) {
 			}
 		}
 
-		<-time.After(1000 * time.Millisecond) // We wait for a bit so the previous log line has time to complete
```
Comment (kerzhner) [inline node/engine/engine.go:866]: Does this line still have a purpose? Does this help us get more debugging information in logs? Maybe @lalexgap or @geoknee remeber?

**Review by kerzhner — APPROVED** [review]

File: node/engine/engine.go (hunk 9 above)
Comment (bitwiseguy) [reply node/engine/engine.go:866]: It looks like a `Panic` level log by zerolog will ensure that the log gets written, then will call `panic()`:
```
		e.logger.Panic().Msg(err.Error())
```

Comment (geoknee) [reply node/engine/engine.go:866]: Let's use that, then :-)

**Pull request merged**